python/ray/serve/_private/application_state.py

```diff
@@ -141,127 +237,191 @@ def deploy(self, deployment_params: List[Dict]) -> List[str]:
+        for params in deployment_params:
+            params["deployment_name"] = params.pop("name")
+            params["app_name"] = self._name
+
+        deployment_infos = {
+            params["deployment_name"]: deploy_args_to_deployment_info(**params)
+            for params in deployment_params
+        }
+        self._set_target_state(deployment_infos=deployment_infos)
+
+    def update_obj_ref(self, deploy_obj_ref: ObjectRef, deployment_time: int) -> None:
+        """Update deploy object ref.
+
+        Updates the deployment timestamp, sets status to DEPLOYING, and
+        wipes the current target state. While the task to build the
+        application is running, the application state stops the
+        reconciliation loop and waits on the task to finish and call
+        apply_deployments_args().
+        """
+
+        if self._deploy_obj_ref:
+            logger.info(
+                f'Received new config for application "{self._name}". '
+                "Cancelling previous request."
+            )
+            ray.cancel(self._deploy_obj_ref)
 
-    def update_obj_ref(self, deploy_obj_ref: ObjectRef, deployment_time: int):
         self._deploy_obj_ref = deploy_obj_ref
         self._deployment_timestamp = deployment_time
-        self._update_status(ApplicationStatus.DEPLOYING)
+        # Halt reconciliation of target deployments
+        self._set_target_state(deployment_infos=None)
 
-    def _process_terminating_deployments(self):
-        """Update the tracking for all deployments being deleted
+    def _get_live_deployments(self) -> List[str]:
+        return self._deployment_state_manager.get_deployments_in_application(self._name)
 
-        When a deployment's status is None, the deployment will be
-        removed from application.
+    def _check_deployment_statuses_and_update_app_status(self) -> bool:
+        """Check deployment statuses, and update the application status correspondingly.
+
+        - If all deployments are healthy, mark the application status
+          as RUNNING
+        - Else if there are unhealthy deployments:
+          - If the current status is DEPLOYING or DEPLOY_FAILED, mark the
+            new application status as DEPLOY_FAILED
+          - Otherwise, mark the new application status as UNHEALTHY
+        - Else (there are updating deployments) don't update the
+          application status
+
+        Returns:
+            A boolean indicating whether the app is ready to be deleted
         """
-        for name in list(self._deployments_to_delete):
-            if self._deployment_state_manager.get_deployment(name):
-                logger.warning(
-                    f"Deleting deployment {name} from application {self._name}."
+
+        # If we're waiting on the build app task to finish, we don't
+        # have info on what the target list of deployments is, so we
+        # can't check on deployment statuses
+        if self._target_state.deployment_infos is None:
```

Comment:
is there is a case when object_ref = None too?

nit: 
if self._deploy_obj_ref:
    return Fasle

Reply:
I reset the deploy_obj_ref to None once we fetch the result from the deploy task, so if deploy_obj_ref=None, it just means that the task has finished.